### PR TITLE
[Enhancement] array_agg support distinct

### LINKED ISSUE #20519: support array_agg(distinct)
sr already support array_agg(), array_agg(order by). Still, we need to support array_agg(distinct)

### PR DESCRIPTION
**Pull request by fzhedu**
Fixes #20519 

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 36 / 37 (97.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|1|2|50.00%|[155]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|29|29|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 15 / 32 (46.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|2|19|10.53%|[35, 37, 38, 50, 54, 57, 58, 61, 63, 64, 65, 66, 67, 68, 75, 77, 86]|
|:large_blue_circle:|src/exprs/agg/factory/aggregate_resolver_avg.cpp|13|13|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 36 / 37 (97.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|1|2|50.00%|[155]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|29|29|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 15 / 32 (46.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|2|19|10.53%|[35, 37, 38, 50, 54, 57, 58, 61, 63, 64, 65, 66, 67, 68, 75, 77, 86]|
|:large_blue_circle:|src/exprs/agg/factory/aggregate_resolver_avg.cpp|13|13|100.00%|[]|

File: be/src/exprs/agg/array_agg.h [hunk 1]
```diff
@@ -2,39 +2,88 @@
...
+    using KeyType = typename SliceHashSet::key_type;
+    void update(MemPool* mem_pool, const ColumnType& column, size_t offset, size_t count) {
+        if constexpr (is_distinct) {
+            if constexpr (pt_is_string<PT>) {
+                for (int i = 0; i < count; i++) {
+                    auto raw_key = column.get_slice(offset + i);
+                    if (set.find(raw_key) == set.end()) {
+                        data_column.append(column, offset + i, 1);
+                        KeyType key(raw_key);
+                        set.template lazy_emplace(key, [&](const auto& ctor) {
```
Comment (kangkaisen) [inline be/src/exprs/agg/array_agg.h:32]: why not directly call lazy_emplace? this way, you call find_or_prepare_insert method twice.

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 36 / 37 (97.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|1|2|50.00%|[155]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|29|29|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 15 / 32 (46.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|2|19|10.53%|[35, 37, 38, 50, 54, 57, 58, 61, 63, 64, 65, 66, 67, 68, 75, 77, 86]|
|:large_blue_circle:|src/exprs/agg/factory/aggregate_resolver_avg.cpp|13|13|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 36 / 45 (80.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|0|3|00.00%|[266, 267, 268]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|5|20.00%|[1675, 1676, 1677, 1679]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|3|4|75.00%|[274]|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|3|4|75.00%|[161]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|28|28|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 15 / 32 (46.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|2|19|10.53%|[35, 37, 38, 50, 54, 57, 58, 61, 63, 64, 65, 66, 67, 68, 75, 77, 86]|
|:large_blue_circle:|src/exprs/agg/factory/aggregate_resolver_avg.cpp|13|13|100.00%|[]|

**Review by satanson — DISMISSED** [review]

**Review by stdpain — DISMISSED** [review]

File: be/src/exprs/agg/array_agg.h (hunk 1 above)
Comment (fzhedu) [reply be/src/exprs/agg/array_agg.h]: refactor it by just emplace at update, and convert set to data column when serialize or finalize.

**Review by stdpain — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 15 / 40 (37.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|2|27|07.41%|[35, 37, 38, 50, 52, 54, 57, 58, 61, 64, 65, 66, 67, 68, 70, 71, 74, 75, 76, 77, 80, 81, 86, 91, 92]|
|:large_blue_circle:|src/exprs/agg/factory/aggregate_resolver_avg.cpp|13|13|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 36 / 45 (80.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|0|3|00.00%|[266, 267, 268]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|5|20.00%|[1675, 1676, 1677, 1679]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|3|4|75.00%|[274]|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|3|4|75.00%|[161]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|28|28|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DecimalV3FunctionAnalyzer.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 15 / 40 (37.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/agg/array_agg.h|2|27|07.41%|[35, 37, 38, 50, 52, 54, 57, 58, 61, 64, 65, 66, 67, 68, 70, 71, 74, 75, 76, 77, 80, 81, 86, 91, 92]|
|:large_blue_circle:|src/exprs/agg/factory/aggregate_resolver_avg.cpp|13|13|100.00%|[]|

**Pull request merged**